File: backend/internal/pkg/status/status.go
package status

// StatusDef defines a status code with its display name and terminal flag.
type StatusDef struct {
	Code     string
	Name     string
	Terminal bool
}

// MainItemStatuses is the single source of truth for main item status definitions.
var MainItemStatuses = map[string]StatusDef{
	"pending":     {Code: "pending", Name: "待开始", Terminal: false},
	"progressing": {Code: "progressing", Name: "进行中", Terminal: false},
	"blocking":    {Code: "blocking", Name: "阻塞中", Terminal: false},
	"pausing":     {Code: "pausing", Name: "已暂停", Terminal: false},
	"reviewing":   {Code: "reviewing", Name: "待验收", Terminal: false},
	"completed":   {Code: "completed", Name: "已完成", Terminal: true},
	"closed":      {Code: "closed", Name: "已关闭", Terminal: true},
}

// SubItemStatuses is the single source of truth for sub item status definitions.
var SubItemStatuses = map[string]StatusDef{
	"pending":     {Code: "pending", Name: "待开始", Terminal: false},
	"progressing": {Code: "progressing", Name: "进行中", Terminal: false},
	"blocking":    {Code: "blocking", Name: "阻塞中", Terminal: false},
	"pausing":     {Code: "pausing", Name: "已暂停", Terminal: false},
	"completed":   {Code: "completed", Name: "已完成", Terminal: true},
	"closed":      {Code: "closed", Name: "已关闭", Terminal: true},
}

// GetMainItemStatus returns the StatusDef for a main item status code.
func GetMainItemStatus(code string) (StatusDef, bool) {
	def, ok := MainItemStatuses[code]
	return def, ok
}

// GetSubItemStatus returns the StatusDef for a sub item status code.
func GetSubItemStatus(code string) (StatusDef, bool) {
	def, ok := SubItemStatuses[code]
	return def, ok
}

// IsMainTerminal returns true if the code is a terminal status for main items.
func IsMainTerminal(code string) bool {
	def, ok := MainItemStatuses[code]
	return ok && def.Terminal
}

// IsSubTerminal returns true if the code is a terminal status for sub items.
func IsSubTerminal(code string) bool {
	def, ok := SubItemStatuses[code]
	return ok && def.Terminal
}

File: frontend/src/lib/status.ts
export const MAIN_ITEM_STATUSES = {
  pending:     { name: '待开始', variant: 'planning',    terminal: false },
  progressing: { name: '进行中', variant: 'in-progress', terminal: false },
  blocking:    { name: '阻塞中', variant: 'overdue',     terminal: false },
  pausing:     { name: '已暂停', variant: 'on-hold',     terminal: false },
  reviewing:   { name: '待验收', variant: 'pending',     terminal: false },
  completed:   { name: '已完成', variant: 'completed',   terminal: true },
  closed:      { name: '已关闭', variant: 'cancelled',   terminal: true },
} as const

export const SUB_ITEM_STATUSES = {
  pending:     { name: '待开始', variant: 'planning',    terminal: false },
  progressing: { name: '进行中', variant: 'in-progress', terminal: false },
  blocking:    { name: '阻塞中', variant: 'overdue',     terminal: false },
  pausing:     { name: '已暂停', variant: 'on-hold',     terminal: false },
  completed:   { name: '已完成', variant: 'completed',   terminal: true },
  closed:      { name: '已关闭', variant: 'cancelled',   terminal: true },
} as const

/** English status codes for filter dropdowns (all main item statuses) */
export const STATUS_OPTIONS = Object.keys(MAIN_ITEM_STATUSES)

/** Lookup variant from either main or sub item status maps */
export function getStatusVariant(status: string): string {
  const def = (MAIN_ITEM_STATUSES as Record<string, { variant: string }>)[status]
    || (SUB_ITEM_STATUSES as Record<string, { variant: string }>)[status]
  return def?.variant ?? 'default'
}

/** Lookup Chinese display name from either main or sub item status maps */
export function getStatusName(status: string): string | undefined {
  const def = (MAIN_ITEM_STATUSES as Record<string, { name: string }>)[status]
    || (SUB_ITEM_STATUSES as Record<string, { name: string }>)[status]
  return def?.name
}

/** Check if an item is overdue based on expected end date and status code */
export function isOverdue(expectedEndDate?: string, status?: string): boolean {
  if (!expectedEndDate || !status) return false
  const isTerminal = (MAIN_ITEM_STATUSES as Record<string, { terminal: boolean }>)[status]?.terminal
    || (SUB_ITEM_STATUSES as Record<string, { terminal: boolean }>)[status]?.terminal
  if (isTerminal) return false
  return new Date(expectedEndDate) < new Date()
}
